Core/clim-basic/extended-streams/stream-input.lisp

```diff
@@ -319,224 +316,70 @@ keys read."))
-      until (or (characterp char) (eq reason :timeout) (eq reason :eof) )
-      finally (return (cond ((eq reason :timeout)
-                             nil)
-                            ((eq reason :eof)
-                             :eof)
-                            (t (char-for-read char)))))))
-
-(defmethod stream-unread-char ((stream standard-extended-input-stream)
-                               char)
-  (with-encapsulating-stream (estream stream)
-    (stream-unread-gesture estream char)))
-
-(defmethod stream-peek-char ((stream standard-extended-input-stream))
-  (with-encapsulating-stream (estream stream)
-    (loop
-      with char and reason
-      do (setf (values char reason) (stream-read-gesture estream :peek-p t))
-      until (or (characterp char) (eq reason :eof))
-      do (stream-read-gesture estream) ; consume pointer gesture
-      finally (return (if (eq reason :eof)
-                          reason
-                          (char-for-read char))))))
-
-(defmethod stream-listen ((stream standard-extended-input-stream))
-  (with-encapsulating-stream (estream stream)
-    (loop
-      with char and reason
-      do (setf (values char reason) (stream-read-gesture estream
-                                                         :timeout 0
-                                                         :peek-p t))
-      until (or (characterp char) (eq reason :eof) (eq reason :timeout))
-      do (stream-read-gesture estream) ; consume pointer gesture
-      finally (return (characterp char)))))
-
-(defmethod stream-clear-input ((stream standard-extended-input-stream))
-  (with-encapsulating-stream (estream stream)
-    (loop
-      with char and reason
-      do (setf (values char reason) (stream-read-gesture estream
-                                                         :timeout 0
-                                                         :peek-p t))
-      until (or (eq reason :eof) (eq reason :timeout))
-      do (stream-read-gesture estream) ; consume pointer gesture
-      ))
-  nil)
+  (loop
+    for ch = (stream-read-gesture stream)
+    when (or (characterp ch) (null ch))
+      return ch))
 
-;;; stream-read-line returns a second value of t if terminated by eof.
-(defmethod stream-read-line ((stream standard-extended-input-stream))
-  (with-encapsulating-stream (estream stream)
-    (let ((result (make-array 1
-                              :element-type 'character
-                              :adjustable t
-                              :fill-pointer 0)))
-      (loop for char = (stream-read-char estream)
-            while (and (characterp char) (not (char= char #\Newline)))
-            do (vector-push-extend char result)
-            finally (return (values (subseq result 0)
-                                    (not (characterp char))))))))
+(defmethod stream-unread-char ((stream standard-extended-input-stream) char)
+  (stream-unread-gesture stream char))
```

Comment:
The method for `standard-input-stream` is almost identical but ensures that `char` is of type `character`. Please add a comment explaining the difference.

Reply:
I don't understand the purpose of adding that comment, it is mentioned in other comments, that SIS is a character stream and that SEIS is not (and methods are in quite different places)

Reply:
I'm not saying the method is wrong. Everything makes sense in terms of the protocols. I just mean a comment like ";; Remember that CHAR is a CHARACTER or an EVENT." could help somebody seeing the method name `stream-unread-CHAR` and the argument name `char` without having studied the relevant parts of the specification or source code recently.

Alternatively, maybe this is another opportunity for `(check-type char input-gesture-obect)`?

Reply:
if anywhere, such check belongs to stream-unread-gesture (added it there). as of the comment, there is already a comment a few lines above

```
;;; FIXME STANDARD-EXTENDED-INPUT-STREAM isn't a character stream, however the
;;; implementation of ACCEPT assumes that it is.
```

if someone (i.e accept) calls this method on something else than a character then it is a violation of the protocol, but for that method being already a hack (which requires fixing), it will accept anything, so suggesting that char may be an event would only add confusion to this method.

Reply:
I don't understand. If `accept` assumes that the stream is a character stream, `accept` would **not** call `stream-unread-char` with non-characters. It could mean that `accept` is unprepared to get non-characters **out of** the stream.

Do I understand correctly that, ultimately, `standard-input-stream` and `extended-input-stream` should/will only accept characters as arguments to `stream-unread-char`?

Reply:
standard-input-stream should only accept characters as arguments to stream-unread-char, yes, and it checks the type of a character. Regarding extended-input-stream, it should not have the method on stream-unread-char and stream-read-char at all, because it is not a fundamental character input stream -- the problem is that some accept methods call read-char on the seis (what is an issue which should be addressed, hence the fixme), and yes, some accept methods are not prepared to get non-characters out of the stream. That's why seis method stream-read-char rejects non-character events to return a character.

Reply:
This is resolved after more discussion in IRC.